omniledger/service/service.go

```diff
@@ -776,20 +811,168 @@ func (s *Service) getTxs(leader *network.ServerIdentity, scID skipchain.SkipBloc
+					log.Error(s.ServerIdentity(), err)
+				}
+			case <-s.heartbeatsClose:
+				log.Lvl2(s.ServerIdentity(), "closing heartbeat timeout monitor")
+				return
+			}
+		}
+	}()
+}
+
+func (s *Service) startViewChange(scID skipchain.SkipBlockID) error {
+	sb, err := s.db().GetLatestByID(scID)
+	if err != nil {
+		return err
+	}
+	if len(sb.Roster.List) < 2 {
+		return errors.New("roster size is too small")
+	}
+	if !sb.Roster.List[1].Equal(s.ServerIdentity()) {
+		// i'm not the next leader, do nothing
+		return nil
+	}
+
+	newRoster := onet.NewRoster(append(sb.Roster.List[1:], sb.Roster.List[0]))
+	genesisDarcID, _, err := s.GetCollectionView(scID).GetValues(GenesisReferenceID.Slice())
+	if err != nil {
+		return err
+	}
+	newRosterBuf, err := protobuf.Encode(newRoster)
+	if err != nil {
+		return err
+	}
+
+	ctx := ClientTransaction{
+		Instructions: []Instruction{{
+			InstanceID: InstanceID{
+				DarcID: genesisDarcID,
+				SubID:  oneSubID,
+			},
+			Nonce:  GenNonce(),
+			Index:  0,
+			Length: 1,
+			Invoke: &Invoke{
+				Command: "view_change",
+				Args: []Argument{{
+					Name:  "roster",
+					Value: newRosterBuf,
+				}},
+			},
+		}},
+	}
+	signer := darc.NewSignerEd25519(s.ServerIdentity().Public, s.getPrivateKey())
+	if err = ctx.Instructions[0].SignBy(signer); err != nil {
+		return err
+	}
+
+	log.Lvlf2("%s: proposing view-change for %x", s.ServerIdentity(), scID)
+	_, err = s.createNewBlock(scID, newRoster, []ClientTransaction{ctx})
+	return err
+}
+
+// getPrivateKey is a hack that creates a temporary TreeNodeInstance and gets
+// the private key out of it. We have to do this because we cannot access the
+// private key from the service.
+func (s *Service) getPrivateKey() kyber.Scalar {
+	tree := onet.NewRoster([]*network.ServerIdentity{s.ServerIdentity()}).GenerateBinaryTree()
+	tni := s.NewTreeNodeInstance(tree, tree.Root, "dummy")
+	return tni.Private()
+}
+
+// withinInterval checks whether public key targetPk in skipchain scID has the
+// right to be the new leader at the current time. This function should only be
+// called when view-change is enabled.
+func (s *Service) withinInterval(scID skipchain.SkipBlockID, targetPk kyber.Point) error {
+	interval, err := s.LoadBlockInterval(scID)
+	if err != nil {
+		return err
+	}
+	t, err := s.heartbeats.getLatestHeartbeat(string(scID))
+	if err != nil {
+		return err
+	}
+
+	// After a leader dies, good nodes wait for 2*interval before accepting
+	// new leader proposals. For every time window of 2*interval
+	// afterwards, the "next" node in the roster list has a chance to
+	// propose to be a new leader.
+
+	currTime := time.Now()
+	if t.Add(time.Duration(rotationWindow) * interval).After(currTime) {
+		return errors.New("not ready to accept new leader yet")
+	}
+
+	// find the position in the proposer queue
+	latestConfig, err := s.LoadConfig(scID)
+	if err != nil {
+		return err
+	}
+	pos := func() int {
+		var ctr int
+		for _, pk := range latestConfig.Roster.Publics() {
```

Comment:
[roster.Search](https://godoc.org/github.com/dedis/onet#Roster.Search) maybe?

Reply:
can't use `roster.Search` because it takes server identities instead of public keys